### PR TITLE
hub/iota: Add local POW

### PR DESCRIPTION
**Pull request by tsvisabo**
# Description

Add support for local Pow, behavior is defined by flag

Fixes # (issue)

## Type of change

New class, flag on server

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?
using api call `broadcastTransactions` and verifying the output

- Sanity tests to see that no future change will truncate txs

# Checklist:

_Please delete items that are not relevant._

- [ ] My code follows the style guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: hub/server/server.cc [hunk 1]
```diff
@@ -57,6 +58,9 @@ DEFINE_string(signingServerChainCert, "/dev/null",
 DEFINE_string(signingServerKeyCert, "/dev/null",
               "Path to SSL certificate key (server.key)");
 
+// remote/local pow settings
+DEFINE_string(powMode, "remote", "pow method to use {remote,local}");
```
Comment (th0br0) [inline hub/server/server.cc]: `PoW method`

File: hub/server/server.cc [hunk 2]
```diff
@@ -103,8 +107,15 @@ void HubServer::initialise() {
...
-  iota::POWManager::get().setProvider(std::make_unique<iota::RemotePOW>(
-      _api, FLAGS_depth, FLAGS_minWeightMagnitude));
+  if (FLAGS_powMode == "remote") {
+    iota::POWManager::get().setProvider(std::make_unique<iota::RemotePOW>(
+        _api, FLAGS_depth, FLAGS_minWeightMagnitude));
+  } else if (FLAGS_powMode == "local") {
+    iota::POWManager::get().setProvider(std::make_unique<iota::LocalPOW>(
+        FLAGS_depth, FLAGS_minWeightMagnitude));
+  } else {
+    LOG(FATAL) << "POW mode: \"" << FLAGS_powMode << "\" not recognized";
```
Comment (th0br0) [inline hub/server/server.cc]: `PoW mode`

File: hub/iota/tests/local_pow.cc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  std::vector<std::string> transactionsTrytes;
+  for (auto i = 0; i < 10; ++i) {
+    transactionsTrytes.push_back(TX_TRYTES);
+  }
+  auto powedTxs = powProvider.performPOW(transactionsTrytes);
+
+  EXPECT_EQ(powedTxs.size(), transactionsTrytes.size());
+  EXPECT_TRUE(
+      std::all_of(powedTxs.begin(), powedTxs.end(),
+                  [](const auto& powedTx) { return powedTx.size() == 2673; }));
```
Comment (th0br0) [inline hub/iota/tests/local_pow.cc]: You should also test, that the various tx form a chain

1 - 2 - 3 - trunk
 \ _ | _  | _ branch

Comment (tsvisabo) [reply hub/iota/tests/local_pow.cc]: Nice drawing!

File: hub/iota/local_pow.cc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+std::vector<std::string> LocalPOW::doPOW(const std::vector<std::string>& trytes,
+                                         const std::string& trunk,
+                                         const std::string& branch) const {
+  auto timestampSeconds =
+      std::chrono::duration_cast<std::chrono::seconds>(
+          std::chrono::system_clock::now().time_since_epoch())
+          .count();
+
+  std::vector<std::string> powedTxs;
+  tryte_t trytesTime[10] = "999999999";
```
[the remark this answers is not in the document]
Comment (tsvisabo) [reply hub/iota/local_pow.cc]: There's '\0'

File: hub/iota/local_pow.cc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  long_to_trytes(timestampSeconds, trytesTime);
+  std::string prevTx;
```
[the remark this answers is not in the document]
Comment (tsvisabo) [reply hub/iota/local_pow.cc]: sure

**Pull request merged**